### PR TITLE
Multi tenant monitoring annotation parsing

### PR DESCRIPTION
**Pull request by gokhangulbiz**
DESCRIPTION: PR description that will go into the change log, up to 78 characters

### CODE REVIEW

File: src/backend/distributed/utils/attribute.c [hunk 1]
```diff
@@ -600,3 +576,35 @@ MultiTenantMonitorshmemSize(void)
...
+		i += 2;
+		while (inputString[i] && (inputString[i] != '*' && inputString[i + 1] != '/'))
+		{
+			i++;
+		}
+	}
+
+	if (i > 2)
+	{
+		char *result = (char *) malloc(sizeof(char) * (i - 1));
```
Comment (marcocitus) [inline src/backend/distributed/utils/attribute.c:602]: should this be palloc?

File: src/backend/distributed/utils/attribute.c [hunk 2]
```diff
@@ -600,3 +577,55 @@ MultiTenantMonitorshmemSize(void)
...
+/*
+ * extractTopComment extracts the top-level multi-line comment from a given input string.
+ */
+static char *
+extractTopComment(const char *inputString)
+{
+	int i = 0;
+
+	/* If query starts with a comment */
+	if (inputString[i] == '/' && inputString[i + 1] == '*')
```
Comment (marcocitus) [inline src/backend/distributed/utils/attribute.c:591]: would this crash if inputString is an empty string?

File: src/backend/distributed/utils/attribute.c [hunk 3]
```diff
@@ -600,3 +575,63 @@ MultiTenantMonitorshmemSize(void)
...
+	}
+
+	int i = 0;
+
+	/* If query starts with a comment */
+	if (inputString[i] == '/' && inputString[i + 1] == '*')
+	{
+		/* Skip the comment start characters */
+		i += 2;
+		while (inputString[i] && (inputString[i] != '*' && inputString[i + 1] != '/'))
```
Comment (marcocitus) [inline src/backend/distributed/utils/attribute.c:599]: do we plan to do some kind of escaping to handle situations like tenant_id = `"akdhek*/a"`?

**Review by halilozanakgul — CHANGES_REQUESTED** [review]

File: src/backend/distributed/utils/attribute.c [hunk 4]
```diff
@@ -619,3 +602,116 @@ MultiTenantMonitorshmemSize(void)
...
+	if (inputStringLen < commentStartCharsLength)
+	{
+		return NULL;
+	}
+
+	int commentEndCharsIndex = 0;
+
+	/* If query starts with a comment */
+	if (inputString[commentEndCharsIndex] == '/' &&
+		inputString[commentEndCharsIndex + 1] == '*')
```
Comment (halilozanakgul) [inline src/backend/distributed/utils/attribute.c:624]: Can you instead make this something like
``` C
if (string[..] != '/' || string[..+1] != '*') return NULL;
```
to reduce nesting

**Review by halilozanakgul — CHANGES_REQUESTED** [review]

File: src/backend/distributed/utils/jsonbutils.c [hunk 5]
```diff
@@ -83,6 +83,25 @@ ExtractFieldBoolean(Datum jsonbDoc, const char *fieldName, bool defaultValue)
...
+ */
+int32
+ExtractFieldInt32(Datum jsonbDoc, const char *fieldName, int32 defaultValue)
+{
+	Datum jsonbDatum = 0;
+	bool found = ExtractFieldJsonb(jsonbDoc, fieldName, &jsonbDatum, false);
+	if (!found)
+	{
+		return defaultValue;
+	}
```
[the remark this answers is not in the document]
Comment (gokhangulbiz) [reply src/backend/distributed/utils/jsonbutils.c:98]: As we discussed in the meeting, we decided to set colocationGroupId to `INVALID_COLOCATION_ID` if `cId` doesn't exist in the annotation.

File: src/backend/distributed/utils/attribute.c [hunk 6]
```diff
@@ -619,3 +602,116 @@ MultiTenantMonitorshmemSize(void)
...
+	{
+		/* Skip the comment start characters */
+		commentEndCharsIndex += commentStartCharsLength;
+		while (inputString[commentEndCharsIndex] &&
```
[the remark this answers is not in the document]
Comment (gokhangulbiz) [reply src/backend/distributed/utils/attribute.c]: My bad. It's unneccessary.

File: src/backend/distributed/utils/attribute.c [hunk 7]
```diff
@@ -619,3 +604,119 @@ MultiTenantMonitorshmemSize(void)
...
+	return substr;
+}
+
+
+/*  EscapeCommentChars adds a backslash before each occurrence of '*' or '/' in the input string */
+static char *
+EscapeCommentChars(const char *str)
+{
+	int originalStringLength = strlen(str);
+	char *escapedString = (char *) malloc(originalStringLength * 2 + 1);
```
Comment (marcocitus) [inline src/backend/distributed/utils/attribute.c:681]: we should basically never use malloc

File: src/backend/distributed/utils/attribute.c [hunk 8]
```diff
@@ -175,60 +182,31 @@ AttributeQueryIfAnnotated(const char *query_string, CmdType commandType)
...
+			colocationGroupId = ExtractFieldInt32(jsonbDatum, "cId",
+												  INVALID_COLOCATION_ID);
 
-		if (MultiTenantMonitoringLogLevel != CITUS_LOG_LEVEL_OFF)
-		{
-			ereport(NOTICE, (errmsg("attributing query to tenant: %s",
-									quote_literal_cstr(tenantId))));
+			if (MultiTenantMonitoringLogLevel != CITUS_LOG_LEVEL_OFF)
+			{
+				ereport(NOTICE, (errmsg(
```
Comment (marcocitus) [inline src/backend/distributed/utils/attribute.c:202]: should this use MultiTenantMonitoringLogLevel instead of NOTICE?

File: src/backend/distributed/utils/attribute.c [hunk 9]
```diff
@@ -619,3 +604,119 @@ MultiTenantMonitorshmemSize(void)
...
+	/* Skip the comment start characters */
+	commentEndCharsIndex += commentStartCharsLength;
+	while (commentEndCharsIndex < inputStringLen &&
+		   !(inputString[commentEndCharsIndex] == '*' &&
+			 inputString [commentEndCharsIndex + 1] == '/'))
+	{
+		commentEndCharsIndex++;
+	}
+
+	if (commentEndCharsIndex > commentStartCharsLength)
```
Comment (marcocitus) [inline src/backend/distributed/utils/attribute.c:640]: for completeness, we should handle the case where there is no */

File: src/backend/distributed/utils/attribute.c [hunk 10]
```diff
@@ -175,60 +182,31 @@ AttributeQueryIfAnnotated(const char *query_string, CmdType commandType)
...
+		if (annotation != NULL)
 		{
-			if (tenantEnd[0] == ' ' && tenantEnd[1] == '*' && tenantEnd[2] == '/')
+			Datum jsonbDatum = DirectFunctionCall1(jsonb_in, PointerGetDatum(annotation));
+
+			text *tenantIdTextP = ExtractFieldTextP(jsonbDatum, "tId");
+			if (tenantIdTextP != NULL)
 			{
-				break;
+				char *tenantId = UnescapeCommentChars(text_to_cstring(tenantIdTextP));
```
[the remark this answers is not in the document]
Comment (marcocitus) [reply src/backend/distributed/utils/attribute.c:193]: consider using postgres functions (e.g. text_substr) as much as possible

unicode is hard

File: src/backend/distributed/utils/attribute.c (hunk 7 above)
Comment (JelteF) [reply src/backend/distributed/utils/attribute.c:681]: for clarity `palloc` is what you want (or any of the other `p` prefixed functions)

File: src/backend/distributed/utils/attribute.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (gokhangulbiz) [reply src/backend/distributed/utils/attribute.c:193]: I give it a try to use `text_substr` (see [34e1daf](https://github.com/citusdata/citus/pull/6753/commits/34e1dafb181327ccc5571c60e74f9d17d318fea6)) but it failed because of multibyte chars handling (see [failed ci job](https://app.circleci.com/pipelines/github/citusdata/citus/31182/workflows/78e9f365-05ba-413c-887e-6efb3ee2f097/jobs/1030454)) So I reverted it. We're safe with our own substring implementation.

File: src/backend/distributed/utils/attribute.c [hunk 11]
```diff
@@ -243,8 +221,15 @@ AnnotateQuery(char *queryString, char *partitionColumn, int colocationId)
 	{
 		return queryString;
 	}
+
+	char *commentCharsEscaped = EscapeCommentChars(partitionColumn);
```
Comment (JelteF) [inline src/backend/distributed/utils/attribute.c:225]: This PR should include tests where this escaping is necessary.

File: src/backend/distributed/utils/attribute.c [hunk 12]
```diff
@@ -619,3 +604,126 @@ MultiTenantMonitorshmemSize(void)
...
+}
+
+
+/*  EscapeCommentChars adds a backslash before each occurrence of '*' or '/' in the input string */
+static char *
+EscapeCommentChars(const char *str)
+{
+	int originalStringLength = strlen(str);
+	char *escapedString = (char *) palloc0((originalStringLength * 2 + 1) * sizeof(char));
+	int escapedStringIndex = 0;
```
Comment (JelteF) [inline src/backend/distributed/utils/attribute.c:689]: Using `StringInfo` is nicer and less error prone than implementing this kind of stuff ourselves.

File: src/backend/distributed/utils/attribute.c [hunk 13]
```diff
@@ -619,3 +604,126 @@ MultiTenantMonitorshmemSize(void)
...
+	/* If query doesn't start with a comment, return NULL */
+	if (inputString[commentEndCharsIndex] != '/' ||
+		inputString[commentEndCharsIndex + 1] != '*')
+	{
+		return NULL;
+	}
+
+	/* Skip the comment start characters */
+	commentEndCharsIndex += commentStartCharsLength;
+	while (commentEndCharsIndex < inputStringLen &&
```
Comment (JelteF) [inline src/backend/distributed/utils/attribute.c:633]: nit: The way this the comment was attached to the lines below made me think on a quick look that the loop was also skipping the comment characters. Adding another comment and/or a newline would have prevented that.

```suggestion
	/* Skip the comment start characters */
	commentEndCharsIndex += commentStartCharsLength;
	
	/* Find the first comment end character */
	while (commentEndCharsIndex < inputStringLen &&
```

File: src/backend/distributed/utils/attribute.c [hunk 14]
```diff
@@ -619,3 +604,126 @@ MultiTenantMonitorshmemSize(void)
 
 	return size;
 }
+
+
+/*
+ * ExtractTopComment extracts the top-level multi-line comment from a given input string.
+ */
+static char *
+ExtractTopComment(const char *inputString)
```
Comment (JelteF) [inline src/backend/distributed/utils/attribute.c:613]: Let's add tests for this function

File: src/backend/distributed/utils/attribute.c [hunk 15]
```diff
@@ -619,3 +604,126 @@ MultiTenantMonitorshmemSize(void)
...
+		return NULL;
+	}
+
+	/* Allocate memory for the substring */
+	char *substr = (char *) palloc((end - start + 1) * sizeof(char));
+
+	/* Copy the substring to the new memory location */
+	strncpy_s(substr, end - start + 1, str + start, end - start);
+
+	/* Add null terminator to end the substring */
```
Comment (JelteF) [inline src/backend/distributed/utils/attribute.c:676]:
```suggestion
	/* Add null terminator to end of the substring */
```

File: src/backend/distributed/utils/attribute.c [hunk 16]
```diff
@@ -619,3 +604,126 @@ MultiTenantMonitorshmemSize(void)
...
+	{
+		commentEndCharsIndex++;
+	}
+
+	/* If there is no end of comment chars , return NULL */
+	if (inputString[commentEndCharsIndex] != '*' &&
+		inputString[commentEndCharsIndex + 1] != '/')
+	{
+		return NULL;
+	}
```
Comment (JelteF) [inline src/backend/distributed/utils/attribute.c:645]: This basically reimplements the C `strstr` function afaict. Let's use that one instead.

File: src/backend/distributed/utils/attribute.c [hunk 17]
```diff
@@ -619,3 +604,126 @@ MultiTenantMonitorshmemSize(void)
...
+	else
+	{
+		return NULL;
+	}
+}
+
+
+/* Extracts a substring from the input string between the specified start and end indices.*/
+static char *
+Substring(const char *str, int start, int end)
```
Comment (JelteF) [inline src/backend/distributed/utils/attribute.c:660]: I think we can replace the usage of this function (or at least a large part of its body) with `pnstrdup`.

**Review by JelteF — CHANGES_REQUESTED** [review]
General approach looks good. I added some comments on replacing custom implementations with stdlib functions.

Also this should have some dedicated tests with some edge cases.

File: src/backend/distributed/utils/attribute.c [hunk 18]
```diff
@@ -619,3 +604,128 @@ MultiTenantMonitorshmemSize(void)
...
+	int originalStringLength = strlen(str);
+	char *escapedString = (char *) palloc0((originalStringLength * 2 + 1) * sizeof(char));
+	int escapedStringIndex = 0;
+
+	for (int originalStringIndex = 0; originalStringIndex < originalStringLength;
+		 originalStringIndex++)
+	{
+		if (str[originalStringIndex] == '*' || str[originalStringIndex] == '/')
+		{
+			escapedString[escapedStringIndex++] = '\\';
```
Comment (JelteF) [inline src/backend/distributed/utils/attribute.c:698]: interesting sidenote, that's maybe worth a comment. The backslash is not the only character that could "escape". Everything Postgres its parsing cares about that the `*/` and `/*` character combinations don't exist. So any character (that's not a `*` or `/`) could be used to escape this pattern in the same way.

File: src/backend/distributed/utils/attribute.c [hunk 19]
```diff
@@ -619,3 +604,128 @@ MultiTenantMonitorshmemSize(void)
...
+EscapeCommentChars(const char *str)
+{
+	int originalStringLength = strlen(str);
+	char *escapedString = (char *) palloc0((originalStringLength * 2 + 1) * sizeof(char));
+	int escapedStringIndex = 0;
+
+	for (int originalStringIndex = 0; originalStringIndex < originalStringLength;
+		 originalStringIndex++)
+	{
+		if (str[originalStringIndex] == '*' || str[originalStringIndex] == '/')
```
Comment (JelteF) [inline src/backend/distributed/utils/attribute.c:696]: I feel like we should also escape the escape character
```suggestion
		if (str[originalStringIndex] == '*' || str[originalStringIndex] == '/' || str[originalStringIndex] == '\\')
```

Comment (JelteF) [reply src/backend/distributed/utils/attribute.c:696]: Actually probably it's probably not needed because of the way we do unescaping. Some tests are definitely needed with a backslash in the tenant ID though.

**Comment by gokhangulbiz** [comment]
Moved to #6796
